### PR TITLE
Feature/audio video streaming services

### PR DESCRIPTION
**Pull request by mrapitis**
:boom: Ready For Review :boom: 

<h2>Summary</h2>

This PR introduces new and updates existing API’s that provide H.264 video and PCM audio streaming services.  Utilizing startH264 and startPCM API methods, developers can open video and audio services and subsequently stream raw H264 video and raw PCM audio to sdlcore.  A brief overview regarding SDL services is described below:

<h2>SDL Services</h2>

Once an application is registered successfully and has established a session, the app has access to services.  After registration, the app may start and stop video and audio services over the established session. By default, applications that are registered receive access to two basic services, RPC and Hybrid.

Below are various services defined by SmartDeviceLink
<ul> 
<li>RPC Service – (automatically started when the app registers) defines the specific APIs sdlcore must perform on behalf of an application , establishes the application listing within sdlcore, performs control functions, and manages HMI.</li>
<li>Hybrid Service -  (automatically started when the app registers) mimics the RPC Service directly, but includes binary data as part of the RPC service</li>
<li>Audio Service (started after a successful startPCM request has been completed)
Sends raw PCM audio data from the device to sdlcore. </li>
<li>H.264 Service (started after a successful startH264 request has been completed)
Sends raw H264 video data from the device to sdlcore. </li>
</ul>

<h3>TODO</h3>
- [X] Add functional hooks for AudioStreaming Service into proxy layers: Protocol, Connection, Base.
- [X] Update existing incomplete developer facing api’s for startH264, startPCM, endH264, end endPCM methods.
- [X] Increase AbstractPacketizer, StreamPacketizer read buffer sizes for optimal performance
- [X] Increase default PipedInputStream buffer size for optimal performance.
- [X] Add method to perform video encoding internally in proxy making a generic opengl surface available for app developers. 
- [X] Update java documentation to detail usage of developer facing video / audio streaming API's below.
```
public boolean startH264(InputStream is) 

public OutputStream startH264() 

public void endH264() 

public boolean startPCM(InputStream is) 

public OutputStream startPCM() 

public void endPCM() 
```

### CODE REVIEW

File: sdl_android_lib/src/com/smartdevicelink/encoder/SdlEncoder.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import android.os.Build;
+import android.view.Surface;
+
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+public class SdlEncoder {
+	
+	// parameters for the encoder
+	private static final String _MIME_TYPE = "video/avc"; // H.264/AVC video
+	// private static final String MIME_TYPE = "video/mp4v-es"; //MPEG4 video
+	private static int _FRAME_RATE = 30;
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/encoder/SdlEncoder.java]: Remove statics, make constants for defaults and member variables with setters to overwrite the defaults

File: sdl_android_lib/src/com/smartdevicelink/encoder/SdlEncoder.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	// encoder state
+	private MediaCodec mEncoder;
+	private PipedOutputStream mOutputStream;
+	
+	// allocate one of these up front so we don't need to do it every time
+	private MediaCodec.BufferInfo mBufferInfo;
+	
+	public SdlEncoder (int frameRate, int iFrameInterval, int width,
+		int height, int bitrate, PipedOutputStream outputStream) {
+		_FRAME_RATE = frameRate;
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/encoder/SdlEncoder.java]: There should be "feedback" or ranges supported for each of the variables such as frameRate, bitrate

File: sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java [hunk 3]
```diff
@@ -2738,125 +2861,371 @@ public void endRPCStream() {
...
+		fTask = null;
+
+		if (pcmServiceEndResponse) {
+			return true;
+		} else {
+			return false;
+		}
+	}
+    
+    public Surface createOpenGLInputSurface(int frameRate, int iFrameInterval, int width,
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java]: References to OpenGL can be moved elsewhere (such as the sdlConn.createOpenGLInputSurface method

File: sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java [hunk 4]
```diff
@@ -2738,125 +2861,371 @@ public void endRPCStream() {
...
-	private void NavServiceEnded() {
-		navServiceResponseReceived = true;
-		navServiceResponse = false;
+	private void NavServiceStartedNACK() {
+		navServiceStartResponseReceived = true;
+		navServiceStartResponse = false;
 	}
 	
-	@SuppressWarnings("unused")
     private void AudioServiceStarted() {
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java]: all the camel case issues

File: sdl_android_lib/src/com/smartdevicelink/streaming/AbstractPacketizer.java [hunk 5]
```diff
@@ -9,12 +9,12 @@
 abstract public class AbstractPacketizer {
 
 	protected IStreamListener _streamListener = null;
-
+	private final static int BUFF_READ_SIZE = 1000000;
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/streaming/AbstractPacketizer.java]: Should be updated based on [protocol spec](https://github.com/smartdevicelink/protocol_spec/issues/4)

File: sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java [hunk 6]
```diff
@@ -2738,125 +2861,371 @@ public void endRPCStream() {
...
-		}
-		if (navServiceResponse) {
-			sdlConn.startStream(is, SessionType.NAV, sdlSession.getSessionId());
-			return true;
+
+		FutureTask<Void> fTask =  createFutureTask(new CallableMethod(2000));
+		ScheduledExecutorService scheduler = createScheduler();
+		scheduler.execute(fTask);
+
+		while (!navServiceStartResponseReceived && !fTask.isDone());
```
Comment (justinjdickow) [inline sdl_android_lib/src/com/smartdevicelink/proxy/SdlProxyBase.java]: Issues should be written to address this blocking in all of the PCM and Video life cycle methods after pulling in this code.

**Comment by mrapitis** [comment]
Closing PR as #237  replaces this request.

**Pull request closed without merging**